added an explicit typecast for Ember Array

mrisher opened the pull request:
This may be not be the best change, but I'm running Ember CLI 4.9.2 and needed to make this change to get past the error on `pushObject()` and enumeration. Is there a better way to fix this? Something got deprecated/changed in 4.x that is making your example (and most of what I find on StackOverflow) not work.

<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

mrisher committed:
added an explicit typecast for Ember Array

This may be not be the best change, but I'm running Ember CLI 4.9.2 and needed to make this change to get past the error on `pushObject()` and enumeration. Is there a better way to fix this? Something got deprecated/changed in 4.x that is making your example (and most of what I find on StackOverflow) not work.

mrisher commented:
This seems like the wrong solution, given the deprecation of `A()` I just learned about on the Discord (https://rfcs.emberjs.com/id/0848-deprecate-array-prototype-extensions/). So maybe we would need to rewrite the functions using `todos` instead of merging this change...

mrisher commented:
Okay, seems like also adding `this.todos = this.todos;    // self-assignment to trigger Tracked` and replacing `pushObject()` with `push` would do the trick and be compliant. You'll also have to replace the filter with `return this.todos.filter((todo) => todo.isCompleted == false);`

NullVoxPopuli commented on files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/ember_interactivity_events_state/index.md:
```diff
@@ -214,6 +214,8 @@ This is because Ember extends JavaScript's Array prototype by default, giving us
 There are dozens of these methods, including `pushObjects()`, `insertAt()`, or `popObject()`, which can be used with any type (not just Objects).
 Ember's [ArrayProxy](https://api.emberjs.com/ember/4.2/classes/ArrayProxy) also gives us more handy methods, like `isAny()`, `findBy()`, and `filterBy()` to make life easier.
 
+**Note:** If you receive a console error around the `pushObject()` function (e.g. `this.todos.pushObject is not a function`) you may need to explicitly typecast your JavaScript Array into an Ember Array using the `A()` method. Add `import { A } from '@ember/array';` at the top of the `todo-data.js` file, and replace `@tracked todos = []` with `@tracked todos = A()`.
```
I think, instead of adding this disclaimer, we should update to using `TrackedArray`, which has regular push <3

NullVoxPopuli commented:
Welcome! No worries! It is newish, and does need to be installed, there is [an RFC](https://github.com/emberjs/rfcs/pull/812) that proposed adding it by default, so i suppose it'll show up by default fairly soon ™


Your contributions mean a lot! <3

Elchi3 commented:
@mrisher Thanks for your PR! Do you want to get back to it and make the changes @NullVoxPopuli proposed?

Elchi3 commented:
@mrisher friendly ping
(This PR might get closed due to inactivity soon)

mrisher commented:
So Sorry, I overlooked this note before and forgot about this. I will check whether I can implement TrackedArray myself in the next few days, if not I'll call for help. Thank you!

mrisher commented:
Hi: I tried to update to use TrackedArray tonight and am hitting bugs; I may not be familiar enough with the Ember internals to debug -- basically, I updated the variable to `todos = new TrackedArray([]);` and removed the self-assignment and everything stopped working with no visible errors. 
Happy to try if someone can help me out, otherwise I think we should retire this PR.

NullVoxPopuli commented:
can you provide the code you had?, the whole file, ideally? <3

mrisher commented:
> 

Sure thing. This is the file adapted from your example code, where `todos` is defined on [line 53](https://github.com/mrisher/junkdrawer/blob/secret-env/app/services/todo-data.js#L53) and then self-assigned on [line 87](https://github.com/mrisher/junkdrawer/blob/secret-env/app/services/todo-data.js#L87) to force the update. Last night I tried changing line 53 to `todos = new TrackedArray([]);` and commenting out the self-assignment (after installing the `tracked-built-ins` package) but then my array stopped updating.

Maybe I need to declare the object type for the TrackedArray?  And IIUC the TrackedArray supports `push()` and `filter()` so I assumed I didn't need to rewrite those, but maybe (I didn't have time to debug further). 

Thanks for the help!

NullVoxPopuli commented:
@mrisher in that repo you provided, does it work without firebase? I don't have an account

mrisher commented:
Ah, I see the problem, it's the assignment when it reads the json blob back from firebase [here](https://github.com/mrisher/junkdrawer/blob/main/app/services/todo-data.js#L213). Is there an idiomatic way to do a deep copy instead of assignment at this line? Or should I rewrite that whole `deserializeJson()` method to take a pointer to the TrackedArray? Thank you for your patience. I'm not aware of an easy way to stub out the firebase code for you...

Elchi3 commented:
@mrisher @NullVoxPopuli sorry, seems like this is stuck (again). I'm closing. Feel free to come back to it, though.
